File: internal/runner/sdk/events.ts
import { appendFileSync, mkdirSync } from "node:fs";
import { createHash, randomBytes } from "node:crypto";

const runID = process.env.OPENFLOW_RUN_ID || "";
const runDir = runID
  ? `${process.env.OPENFLOW_HOME || "~/.openflow"}/runs/${runID}`
  : "";

let runDirReady = false;

function ensureRunDir(): void {
  if (!runDirReady && runDir) {
    mkdirSync(runDir, { recursive: true });
    runDirReady = true;
  }
}

export function emitEvent(event: Record<string, unknown>): void {
  if (!runDir) return;
  try {
    ensureRunDir();
    appendFileSync(`${runDir}/events.jsonl`, JSON.stringify(event) + "\n");
  } catch {
    // best-effort
  }
}

export function log(label: string, ...args: unknown[]): void {
  const msg = args.map(String).join(" ");
  const line = `[openflow] ${label}: ${msg}`;
  process.stderr.write(line + "\n");
}

export function deriveAgentName(systemPrompt: string): string {
  const words = systemPrompt
    .toLowerCase()
    .replace(/[^a-z0-9 ]/g, " ")
    .split(/\s+/)
    .filter((w) => w.length > 0)
    .slice(0, 4)
    .join("-")
    .replace(/^-|-$/g, "");

  const hash = createHash("md5")
    .update(systemPrompt)
    .digest("hex")
    .slice(0, 8);

  return `${words}-${hash}`;
}

export function generateSessionID(): string {
  const now = new Date();
  const ts =
    now.getFullYear().toString() +
    String(now.getMonth() + 1).padStart(2, "0") +
    String(now.getDate()).padStart(2, "0") +
    "_" +
    String(now.getHours()).padStart(2, "0") +
    String(now.getMinutes()).padStart(2, "0") +
    String(now.getSeconds()).padStart(2, "0");
  const suffix = randomBytes(4).toString("hex");
  return `openflow_session_${ts}_${suffix}`;
}

File: internal/runner/sdk/events.go
package sdk

import (
	"encoding/json"
	"os"
	"path/filepath"
	"sync"
)

var (
	runDir      string
	runDirOnce  sync.Once
	runDirReady bool
)

func ensureRunDir() {
	runDirOnce.Do(func() {
		runID := os.Getenv("OPENFLOW_RUN_ID")
		if runID == "" {
			return
		}
		home := os.Getenv("OPENFLOW_HOME")
		if home == "" {
			home, _ = os.UserHomeDir()
			home = filepath.Join(home, ".openflow")
		}
		runDir = filepath.Join(home, "runs", runID)
		os.MkdirAll(runDir, 0755)
		runDirReady = true
	})
}

type Event struct {
	Type       string `json:"type"`
	Agent      string `json:"agent,omitempty"`
	Status     string `json:"status,omitempty"`
	Step       string `json:"step,omitempty"`
	Prompt     string `json:"prompt,omitempty"`
	Resume     bool   `json:"resume,omitempty"`
	Result     string `json:"result,omitempty"`
	Error      string `json:"error,omitempty"`
	Cmd        string `json:"cmd,omitempty"`
	Stdout     string `json:"stdout,omitempty"`
	Stderr     string `json:"stderr,omitempty"`
	ExitCode   int    `json:"exit_code,omitempty"`
	DurationMs int64  `json:"duration_ms,omitempty"`
	Timestamp  string `json:"timestamp"`
}

func emitEvent(e Event) {
	ensureRunDir()
	if runDir == "" || !runDirReady {
		return
	}
	data, err := json.Marshal(e)
	if err != nil {
		return
	}
	f, err := os.OpenFile(filepath.Join(runDir, "events.jsonl"), os.O_APPEND|os.O_CREATE|os.O_WRONLY, 0644)
	if err != nil {
		return
	}
	defer f.Close()
	f.Write(append(data, '\n'))
}
